Source/WebCore/platform/graphics/SourceBufferPrivate.cpp

```diff
@@ -408,6 +408,22 @@ MediaTime SourceBufferPrivate::findPreviousSyncSamplePresentationTime(const Medi
 }
 
 Ref<MediaPromise> SourceBufferPrivate::removeCodedFrames(const MediaTime& start, const MediaTime& end, const MediaTime& currentTime)
+{
+    MediaPromise::Producer producer;
+    Ref<MediaPromise> promise = producer.promise();
+
+    m_currentSourceBufferOperation = m_currentSourceBufferOperation->whenSettled(RunLoop::current(), [weakThis = WeakPtr { *this }, this, start, end, currentTime](auto&& result) mutable {
+        if (!weakThis || !result)
+            return MediaPromise::createAndReject(!result ? result.error() : PlatformMediaError::BufferRemoved);
+        return removeCodedFramesInternal(start, end, currentTime);
+    })->whenSettled(RunLoop::current(), [producer = WTFMove(producer)](auto&& result) mutable {
+        producer.settle(result);
+        return MediaPromise::createAndSettle(WTFMove(result));
+    });
+    return promise;
```

Comment:
Can we return m_currentSourceBufferOperation instead, it seems equivalent to a new promise that gets settled when m_currentSourceBufferOperation gets settled.

Reply:
I would need to make MediaPromise non exclusive. 

I hesitated in doing so, but knowing your reluctance on even having a non exclusive mode. I decided against it. 

With a non exclusive promise we could indeed make the code simpler.

Reply:
I thought of a solution, which I think is quite elegant https://bugs.webkit.org/show_bug.cgi?id=265096